components/script/dom/readablestreamdefaultcontroller.rs

```diff
@@ -2,18 +2,124 @@
+use std::cell::{Cell, RefCell};
+use std::rc::Rc;
+
 use dom_struct::dom_struct;
 use js::rust::HandleValue as SafeHandleValue;
 
+use crate::dom::bindings::cell::DomRefCell;
 use crate::dom::bindings::codegen::Bindings::ReadableStreamDefaultControllerBinding::ReadableStreamDefaultControllerMethods;
+use crate::dom::bindings::codegen::Bindings::UnderlyingSourceBinding::UnderlyingSource as JsUnderlyingSource;
 use crate::dom::bindings::import::module::{Error, Fallible};
-use crate::dom::bindings::reflector::Reflector;
+use crate::dom::bindings::reflector::{reflect_dom_object, Reflector};
+use crate::dom::bindings::root::{DomRoot, MutNullableDom};
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::readablestream::ReadableStream;
+use crate::dom::readablestreamdefaultreader::ReadRequest;
 use crate::script_runtime::JSContext as SafeJSContext;
 
+#[derive(JSTraceable)]
+pub enum UnderlyingSource {
+    /// Facilitate partial integration with sources
+    /// that are currently read into memory.
+    Memory(usize),
+    /// A blob as underlying source, with a known total size.
+    Blob(usize),
+    /// A fetch response as underlying source.
+    FetchResponse,
+    /// A JS object as underlying source.
+    Js(JsUnderlyingSource),
+}
+
+impl UnderlyingSource {
+    /// Is the stream backed by a Rust native source?
+    pub fn is_native(&self) -> bool {
+        !matches!(self, UnderlyingSource::Js(_))
+    }
+
+    /// Does the stream have all data in memory?
+    pub fn in_memory(&self) -> bool {
+        matches!(self, UnderlyingSource::Memory(_))
+    }
+}
+
 /// <https://streams.spec.whatwg.org/#readablestreamdefaultcontroller>
 #[dom_struct]
 pub struct ReadableStreamDefaultController {
     reflector_: Reflector,
+
+    /// <https://streams.spec.whatwg.org/#readablestreamdefaultcontroller-queue>
+    buffer: RefCell<Vec<u8>>,
+
+    #[ignore_malloc_size_of = "Rc is hard"]
+    underlying_source: Rc<UnderlyingSource>,
+
+    stream: MutNullableDom<ReadableStream>,
+}
+
+impl ReadableStreamDefaultController {
+    fn new_inherited(underlying_source: Rc<UnderlyingSource>) -> ReadableStreamDefaultController {
+        ReadableStreamDefaultController {
+            reflector_: Reflector::new(),
+            buffer: RefCell::new(vec![]),
+            stream: MutNullableDom::new(None),
+            underlying_source,
+        }
+    }
+    pub fn new(
+        global: &GlobalScope,
+        underlying_source: Rc<UnderlyingSource>,
+    ) -> DomRoot<ReadableStreamDefaultController> {
+        reflect_dom_object(
+            Box::new(ReadableStreamDefaultController::new_inherited(
+                underlying_source,
+            )),
+            global,
+        )
+    }
+
+    pub fn set_stream(&self, stream: &ReadableStream) {
+        self.stream.set(Some(stream))
+    }
+
+    /// <https://streams.spec.whatwg.org/#ref-for-abstract-opdef-readablestreamcontroller-pullsteps>
+    pub fn perform_pull_steps(&self, read_request: ReadRequest) {
```

Comment:
My issue is more about [`pullAlgorithm`](https://streams.spec.whatwg.org/#readablestreamdefaultcontroller-pullalgorithm) and similar internal slots. Because if you search `Let pullAlgorithm`, you can see there are multiple ways to create it. Do you have any suggestions on how to implement it?

Reply:
Using a trait object or function pointer would be my first attempt.

Reply:
I've added a commit that calls into the pull algorithm.

Reply:
Ah I guess that means we will need to implement the algos to other vairants.